File: math/53.MaximumSubarray.go
package main

import "fmt"

func maxSubArray(nums []int) int {
	ans := nums[0]
	sum := nums[0]
	for _, num := range nums[1:] {
		ans = max(num+ans, num)
		if ans > sum {
			sum = ans
		}
	}
	return sum
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	nums := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	m := maxSubArray(nums)
	fmt.Println("m:", m)
}

File: math/53.MaximumSubarray.py
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        ans = nums[0]
        result = nums[0]
        for num in nums[1:]:
            result = max(result + num, num)
            if result > ans:
                ans = result
        return ans


if __name__ == "__main__":
    s = Solution()
    case1 = [-2, 1, -3, 4, -1, 2, 1, -5, 4]  # 6
    print(s.maxSubArray(case1))
